doc/design/persistent_volumes_etcd_data.md

```diff
@@ -0,0 +1,79 @@
+# Use Persistent Volumes for etcd data
+
+## Proposal and Motivation
+
+Today etcd-operator creates ephemeral etcd members: in their pod the etcd data is stored inside a volume of type emptyDir and the pod restart policy is set to `Never`.
+
+So, if a pod is deleted from the API server (manually or automatically: i.e a node goes down and the node controller, after a timeout, evicts pods marking them for termination), etcd-operator will remove the related etcd member from the etcd cluster and will create a new pod (with a new name and an empty data dir) and add that new etcd member to the etcd cluster.
+
+When a majority of pods is deleted at the same time the etcd cluster will become unquorate and there'll be no way to restore its quorum since the members and their data are gone. So the unique solution (that etcd-operator currently do) is to reinitialize a new cluster restoring its data from a backup (if available).
+
+This proposal is to fix such case achieving a persistent/durable etcd cluster.
+
+## Proposed Changes
+
+#### Changes
+
+Add an option to enable storing etcd data inside a persistent volume and related reconcile logic changes.
+
+The change to be effective and cover the above failure scenario is split in two parts.
+
+### Part 1
+
+- Add to the cluster spec PodPolicy an option to enable putting etcd data inside a Persistent Volume
+- If podPolicy has persistent volumes enabled, when creating a pod also create a Persistent Volume Claim and use it as the volume source for etcd-data
+- Handle etcd data persistent volume garbage collection like already done with other resources. While backup pv aren't garbage collected, etcd data pv needs to be garbage collected when not needed anymore.
+
+#### Pod name to Persistent Volume Claim name mapping
+
+The pod and related PVC are mapped 1 to 1 so thet can be easily generated and deducted. I.E. if the pod is called `etcd-0000` its related PVC name will be `etcd-0000-pvc`
+
+#### Results
+
+These changes will put the basis for initial persistent volume management BUT won't fix the failure scenario described at the start of the proposal: if a majority of pods is deleted there's no logic to recreate them with their previous data.
```

Comment:
> if a majority of pods is deleted there's no logic to recreate them with their previous data.

Why need to recreate the cluster?

I think for the first phase we will just wait the failed etcd pods to restart until they form healthy quorum again.

Reply:
but the pods can be deleted (removed from the API) by k8s itself (or by the user) regardless of restart policy (which is a pod property) as explained previously in the doc.

Reply:
> For this phase, we assume that we will wait for any failed etcd pod to restart instead of recreating it.

This is handled by part one, but as said pods can be removed from the API and Part 2 aims to recreate them.

Reply:
> but the pods can be deleted (removed from the API) by k8s itself (or by the user) regardless of restart policy (which is a pod property) as explained previously in the doc.

Yeah. That's definitely something we will handle, but not in phase/part 1.
I see our divergence here. My reasoning is that we want to do good engineering -- that means solve problems gradually and one by one. It's easier to implement, review, and keep on track. Again, general rule we apply to engineering.